### ISSUE TITLE
NPC deals far too much damage

### ISSUE THREAD

**Issue by Blamo27** [mcmonkeyprojects/Sentinel#122]
Hello,

As you can see on the video below, the NPC (all the sentinels in general, actually) seems to deal A LOT more damage than it is supposed to. Unless we don't really get how the damage system works compared to the one there was on Sentry, a damage of 18 shouldn't take more than 4 hearts to a player fully equipped with a Protection 4 diamond armor. In fact, previously it was equal to about 0.75 heart, so far less.

https://streamable.com/yji98

Also, for testing purposes we put something up to display to the player a message each times he receives damages. That message is "Dur chestplate: X". As you can see, the message displays twice every time the sentinel hits the player, which would imply the sentinel deals twice the damage (previously, with Sentry I mean, there was only one message displayed each time a Sentry attacked). So I think there is a double problem here : not only the Sentinel deals more damage that it is supposed to, but this damage is also inflicted twice (it can't be only that last thing, since a damage of 18 normally takes less than one heart, so the double shouldn't take four hearts).

**Comment by mcmonkey4eva** [comment]
What's your Sentinel config?

**Comment by Blamo27** [comment]
http://pastebin.com/0cKduHrb

**Comment by mcmonkey4eva** [comment]
The relevant damage code is more or less just a Bukkit call: `LivingEntity#Damage(double, Entity)`
Do note that 18 damage, as far as Minecraft, is /a lot/. That's at the least as much as a strongly enchanted diamond sword. It's possible Sentry weakened the damage it output for some reason...

http://minecraft.gamepedia.com/Sword a diamond sword by default does 7 damage. You've nearly tripled that!

I glanced at the old Sentry code and can't even figure out where it applies damage at all so I dunno if that is what happened. It does appear to force-set damage event damage to the strength value, but beyond that I don't know what it's even... doing.

**Comment by Blamo27** [comment]
Actually, it appears that a 18 damage value is exactly equivalent to a sharpness 21 sword. So, we enchanted a sword with sharpness 21 and tried it repeatedly on a player equipped with a full Protection 4 diamond armor. The result is exactly as it were before : a little less than one heart damage (to be exact, it deals 1.963 damage, so roughly 98% of one heart). We did some more tests and try to find what Sharpness value it takes to deal the kind of damage the NPC dealt. The result is baffling : in one shot, the NPC takes 8.443 heart away, which is the roughly the same as a sword enchanted Sharpness... 83 !
So yeah, even by Minecraft standards, 18 should be far, far, far from this kind of damage. Again, it's a player equipped with a full Protection 4 diamond armor we're talking about !

Also, the fact that the damage event seems to happen twice is certainly not normal !

**Comment by mcmonkey4eva** [comment]
Huh. 8.44 is about 9 hearts... which is about 18 damage.
Hah.
Hahah.
Uh.
I wouldn't know offhand why armor just casually doesn't count anymore tbh. Again: I'm literally calling proper bukkit damage method. Everything that happens beyond that is Bukkit and minecraft's fault!

I'll look into alternative methods of apply damage. Though, one is already available: The workaround damage config option. Perhaps that'll yield ... different results?

**Comment by mcmonkey4eva** [comment]
asgdklsagk;lgsa I just looked and it's caused by `enforce damage`, which you have enabled >:(

**Comment by Blamo27** [comment]
Well, obviously the 8.44 damage equals to 4.22 hearts, as shown on the video ! Absolutely not related to 18 damage x)

As for the enforce damage, of course we enabled it, you created it specifically to deal with one of our other issues ! https://github.com/mcmonkey4eva/Sentinel/issues/95

We can't disable it without having this other issue appearing again !

**Comment by mcmonkey4eva** [comment]
`Well, obviously the 8.44 damage equals to 4.22 hearts, as shown on the video ! Absolutely not related to 18 damage x)` you said 8.44 hearts not 8.44 damage.  D:

**Comment by mcmonkey4eva** [comment]
AAaannnnd I don't know what to do then... short of internally calculating all damage modifiers (armor, etc) my self... 


Actually, I have an idea, maybe Spigot will do something properly for once...

**Comment by mcmonkey4eva** [comment]
No promises but new build might fix it.

**Comment by Blamo27** [comment]
You're right, I meant 8.44 hearts in the way that players have 20 hp. It actually takes 8.44 out of 20 hp to the player.

**Comment by Blamo27** [comment]
So, latest build improved things : the damage is now far more reasonable, but is still doubled : by that, I mean it's twice the damage it should be, and the damage event appears to occur twice, as our displayed damage message indicates.

In other words, I think you fixed the first issue I mentionned in my post, but the second one is still there !

**Comment by mcmonkey4eva** [comment]
There's no reason I can see within Sentinel code that it should double - it might be Bukkit is doubling it...?

**Comment by Blamo27** [comment]
You may check the instances of your plugin, is the EntityDamage event set with the right priority ?

**Comment by mcmonkey4eva** [comment]
Why would that change anything 0.o

**Comment by Blamo27** [comment]
Did you test the event by debugging it ?
Make sure you are instantiating the class only once.

I had exactly the same issue on a plugin of mine, but I don't remember why or how...

**Comment by mcmonkey4eva** [comment]
After testing replication:

This only happens when you enable `enforce damage` but not `workaround damage`.
So just flick the workaround config option on and it should be fine.

The basics of why:
The regular damage goes through with complete disregard for the event being cancelled for some reason.
So both the regular damage and the enforced damage apply, causing an exact doubling of damage.
I Suspect a Spigot-side screwup, but don't feel much like bugging them about it.

**Comment by Blamo27** [comment]
Well this is annoying : when we switch workaround damage to true, the damages are indeed not doubled anymore... But they are once again far too high ! In other words, the issue we fixed earlier in this topic comes back when we enable workaround damage, which means the NPC deals much more damage than what it should deals, but they are dealt only one time, thus fixing the second issue. And if we disable the workaround damage, the "too much damage dealt" issue disappears, but the "damage inflicted twice" comes back.
Seems like a dead end... What can we do ?

**Comment by mcmonkey4eva** [comment]
I don't even know tbh.

**Comment by AdraliK** [comment]
Good afternoon. I encountered this problem on my server, but it's a bit different. When I put 15 damage from a NPS with a "player" type, then I lose 2.5 hearts (in diamond armor) when I hit. If I change to a "zombie" type or another hostile mob with the same parameters, I lose 4 hearts . This bug was in all versions of this plugin. I ask that you would fix it in the next update or tell me how to fix it?

**Comment by Blamo27** [comment]
@mcmonkey4eva 
We're still having this issue !
Just in case, have you tried debugging the tryAttack method in the SentinelTrait.java file, to check if it was fired twice when workaround damage is set to false ?

**Comment by mcmonkey4eva** [comment]
I can't see anything happening wrong logically in current version (DEV-132)
as best I can tell, workaround damage is doing the correct damage values, enforce damage is doing double damage for zero visible reason (I debugged the exact values my plugin sent out, but my hearts went down double).

Feel free to look at the source code here on github if you think you can spot something I missed.

Either way, try DEV-132+ from http://ci.citizensnpcs.co/job/Sentinel/
and enable `/sentinel debug`
it will fill your console with what the NPCs are doing.
Should something look wrong... post it here.
If nothing looks wrong, but you still get wrong values in both modes...
Well, I'm open to suggestions? lol

**Comment by Blamo27** [comment]
Sentinel debug did not provide any useful information, the tryAttack seems to occur only once, but the damages are in fact inflicted twice.

Just to clear things up, here is a recap of the issue : 
* Workaround Damage = false ; enforce damage = true   -> Normal damages, but inflicted twice each hit (sentinel with damage value of 18 = 1.963 HP taken away each hit, but since there are two hits in one, it's actually 3.926 HP taken away each time the sentinel strikes).
* Workaround Damage = true ; enforce damage = true   -> Damages inflicted only once, but abnormally high (sentinel with damage value of 18 = 8.443 HP taken away in one hit).
* Workaround Damage = false ; enforce damage = false   -> Damages inflicted only once, with perfectly normal values (sentinel with damage value of 18 = 1.963 HP taken away in one hit, which is exactly what it should would be great of course.

**Comment by mcmonkey4eva** [comment]
Does the log show damage being inflicted twice?
I logged where it applied damage in the debug system...

**Comment by Blamo27** [comment]
(I edited my previous message, i forgot to post a part of what I wrote)

Here is the log of the sentinel debug : as you can see, it first detects the player, goes toward him, attack him, then the player leaves and so does the sentinel. The tryAttack passed only one time but the damages were inflicted twice.
https://pastebin.com/1kbLFXhk

**Comment by mcmonkey4eva** [comment]
... Wait, are you using archers?
I've been testing with fist-punching... >.>

**Comment by Ferocimo** [comment]
Hello, I'm co-founder of the server of Blamo27. I will answer when he can't.

No, this test was performed on a standard sentinel wielding a diamond sword.

**Comment by mcmonkey4eva** [comment]
I don't even see it calculating damage values in the debug log...
As best I'm aware, bar me missing something in the code, it should post damage tracking to the log...

**Comment by Ferocimo** [comment]
Well I don't see what I can do, I enabled sentinel debug on the sentinel and this showed up in the logs, nothing more...
Do you think it might be because the NPC is wielding a sword and not using its own fists ?

**Comment by Ferocimo** [comment]
I think you uploaded the wrong version : we downloaded the version you just sent, but it appears it's the build 131. We tried anyway with the debug but there was no changes at all in the log (same log as what we sent earlier).

**Comment by mcmonkey4eva** [comment]
...
Mate.
That's an automated build server.
It does not mess up uploading files.
You messed up the downloading of it.

**Comment by mcmonkey4eva** [comment]
Wait, you were on 131? You never even ran 132 as I asked?! D:

**Comment by Blamo27** [comment]
No no we were testing the latest version !

**Comment by Blamo27** [comment]
Firstly, you can't put Bukkit events into SentinelTrait Class, because it will fire events a lot ...

**Comment by mcmonkey4eva** [comment]
... what?
That's intentional placement for careful logical reasons...

**Comment by mcmonkey4eva** [comment]
Also, I'm still awaiting an updated log from DEV-138+

**Comment by Blamo27** [comment]
So if you can't move your events out of SentinelTrait Class you should check the number of instances of it.

**Comment by mcmonkey4eva** [comment]
Those events output debug log info if they do anything.
Thus the request for debug logs...

**Comment by Blamo27** [comment]
This is the debug log : https://pastebin.com/shiiXYX8

**Comment by mcmonkey4eva** [comment]
I see it striking once, not twice, in that log...

**Comment by Ferocimo** [comment]
Yes that's what we said. The log doesn't show anything relevant, but the damage is indeed inflicted twice, there's no doubt. Our plugin clearly shows that the player is hit twice at the same time, and takes twice the damage. That only happens when enforce damage is set to true and workaround damage set to false, otherwise the damage is inflicted only once.

**Comment by mcmonkey4eva** [comment]
Your plugin helpfully stated ` 29.08 21:38:04 [Server] INFO [Broadcast] dmg : 2.462400436401367 ` once, not twice :P 
(I caught it before you edited the paste lol)

**Comment by Ferocimo** [comment]
This is not our plugin, that's actually a broadcast we added in your plugin, we removed it from the pastebin precisely to avoid any confusion.

Our plugin shows the damage being inflicted twice directly in the tchat of the player being hit.

**Comment by mcmonkey4eva** [comment]
Are you sure it's your not plugin reading it twice by mistake?
Cause Sentinel is showing it being applied once pretty clearly...

**Comment by Ferocimo** [comment]
As you can see in this video we made when we created this topic, our plugin shows the damage being inflicted twice, with the message "dur chestplate: [x]" displayed twice per hit.
https://streamable.com/yji98

Yes we're sure it's not our plugin reading it twice because the damage are indeed twice as high as what they should be, and what they are when enforce damage is disabled.

**Comment by mcmonkey4eva** [comment]
Can I see source code of your plugin?

**Comment by Ferocimo** [comment]
Blamo doesn't have it where he is right now, however that doesn't really matter. If you want, we can show you a video making a comparison between when we enable enforce damage and when we disable it. When it is disabled, the "dur chestplate" message is only displayed once and the damage is twice as low.

The fact that the damage is doubled is really indisputable.

**Comment by mcmonkey4eva** [comment]
No. Get source code.
Cause I'm betting y'all forget the `isCancelled` check in the event.
And thus are posting messages from cancelled events alongside actual ones.

**Comment by Ferocimo** [comment]
How do you explain the damage being twice as high then ?

**Comment by Ferocimo** [comment]
(we just performerd the tests without our plugin, and it still clearly show that the damage is twice as high as what it is when enforce damage is disabled)

**Comment by mcmonkey4eva** [comment]
First guess is "hearts vs. HP" mixup
Second guess is "uncancellation" by terribly dumb code in either your plugin or one you have on your server.

I'm not trying to be hostile here, but my every attempt at replicating your issue, or debugging it ON YOUR SERVER, has lead to "yeah Sentinel's operating fine" and I'm getting tired of the constant blame being put on Sentinel when there's zero evidence it's done anything wrong and a fair chance something on your server is breaking it.
I won't mind if I'm proven wrong, but please hold back the "It's indisputably [this or that]" until we have some actual facts here.

The only facts I have are those granted by the debug logs, which clearly show damage is not being sent out OR monitored as input twice.

**Comment by Ferocimo** [comment]
Okay that's fair, but it looks like we can't move much further in this issue just with the data we can provide. In that case please perform the test on your personal server and tell us the results : 

First, make a sentinel with 18 damage which will attack the player.
Make sure you have a configuration similar to ours, with workaround damage set to false and enforce damage set to true.
Get hit by the NPC. 
If you have Essentials, use the command /whois to see how much HP you have (like, 16.238 on 20).
Heal yourself and try repeating that multiple times to get a good idea of how much damage you take each time you're hit by the NPC.
Then, set the "enforce damage" option to false in your configuration and repeat the tests. You should see a pretty clear difference in how much health you lose at each hit. If the difference isn't clear enough, you can also, I guess, set the damages of the NPC to a higher value like 36. If it is doubled with this value you should lose nearly 4 hearts (= 8 HP). 

Please tell us if it confirms what we're describing.

**Comment by Ferocimo** [comment]
(our current configuration, if needed : 
https://pastebin.com/Fnq3GBrS )

**Comment by mcmonkey4eva** [comment]
18 damage = 2 hit kill
36 damage = 1 hit kill
...

**Comment by Ferocimo** [comment]
I forgot to mention, you have to wear a Protection IV armor.

**Comment by mcmonkey4eva** [comment]
**Growl**. Something IS uncancelling it... I know because I can trace it down on my test server. Lemme try to workaround it.

**Comment by Ferocimo** [comment]
Okay.

By the way if you want to replicate it again, I just made a more simple test.
I went unequipped (no armor) against the sentinel, and set its damage value to 2. With enforce damage disabled I lose exactly one heart (2 HP) each hit. When it's enabled I lose exactly 2 hearts (4 HP) each hit.
It's clearer this way than with protection IV armor and stuff that modifies the damage you receive.

**Comment by mcmonkey4eva** [comment]
Build DEV-140 applies a workaround to the fact that the cancel is being blocked.

Now I just need to figure out what dev to throttle here (for uncancelling my cancelled event!)...

**Comment by Ferocimo** [comment]
It indeed now inflicts normal damage, which will allow us to continue our work with your plugin. However, as you know, the hit is still technically inflicted twice (as our plugin shows it with the double "dur chestplate" message), it's just that the second hit deals no damage. x)

Please let us know if you find out a cleaner fix to this issue, and thank you for your time and the workaround.

**Comment by Blamo27** [comment]
Also, just to know but, why is EntityDamageByEntityEvent fired at multiple times ?

**Comment by mcmonkey4eva** [comment]
It's fired twice because that's how the enforcement system works :P 
It listens to the damage event, then cancels that event, and "enforces" it by sending a sourceless damage event, which won't be cancelled.

The issue appears to be that if you listen to damage events in certain other plugins, the cancelling goes away.

I'm currently investigating the possibility that there might be an error in Denizen leading to this,
but as best I can tell Denizen isn't touching the cancelled state before the cancel disappears...

**Comment by Ferocimo** [comment]
(just passing though to mention, we do not have Denizen)

**Comment by mcmonkey4eva** [comment]
Then it's probably a spigot issue with "just any plugin listening in"

I can actively cause it to happen or not happen by listening or not listening to the damage event in Denizen

Freakin'... weird spigot bugs :/

**Comment by mcmonkey4eva** [comment]
It's been quite a while ... is this still an issue on current Sentinel versions (1.8.5+?)

**Comment by Ferocimo** [comment]
@mcmonkey4eva No it's still the same. With enforce damage set to true (which we have to use to fix other issues previously reported), archers for instance completely ignore armor : "11 damage" will deal 11 damage regardless of your armor or enchants. 

We still have to test the "two damage events" thing but our developper is currently unavailable.

**Comment by mcmonkey4eva** [comment]
*Is* enforce damage still needed? If it's from WorldGuard - they recently pushed additional patches to improve NPC support in WG 7.

**Comment by Ferocimo** [comment]
I can't test this as we are not yet in 1.13 (but soon will be), required for WG 7.

I will let you know when we will be able to test this.

**Comment by Ferocimo** [comment]
@mcmonkey4eva We are now on 1.13. This is what we found : 

WITHOUT enforce damage : 
Damage values on sentinels are irrelevant. No matter what we put, the NPC will deal 7 damages if it has a diamond sword, less if it has another type of sword, etc. Even if we put 60 damages.

WITH enforce damage : 
NPCs deal the damage of the sword like above (so, 7 if equipped with diamond sword), but also some damages based on what we put as damage value. So two sets of damages are dealt with each hit (which can be clearly seen if you have a plugin that indicates how much damage you receive, like Combat Indicator) : 
-7 (sword damage)
-3.4 (sentinel damage)
With each hit, with a plugin like Combat Indicator you will see these two damages show up.

**Comment by Ferocimo** [comment]
Adding a screenshot so you could see what I mean by "two sets of damages are dealt with each hit" : http://www.noelshack.com/2019-18-6-1556970089-2019-05-04-13-41-14.png 

This is only one NPC attacking here, and he deals 7 damages (base diamond sword damage) AND 10.2 damages (damages scaled to what strength I gave him with the command /sentinel damage).

**Comment by mcmonkey4eva** [comment]
I don't know what to tell you other than that I cannot replicate any of your results in a clean environment.
With `enforce damage` set to `false`, damage is exactly correct as set (set to 15 damage, does 15 damage), and the `EntityDamageEvent` fires once, as expected.

With `enforce damage` set to `true`, damage is exactly correct as set (15 damage does 15 damage), and the `EntityDamageEvent` fires twice in the expected manner: once with the real damage, and once with damage of `0`. There seems to be some inaccuracy with `isCancelled` tracking, but that's fine as the damage is clearly set to `0` still.
The only way to get seemingly doubled damage is to listen to the damage event with priority set to `lowest` to intentionally pick it up before Sentinel does (in priority `low`).

Also I looked up `Combat Indicator` on SpigotMC to check if it's doing anything relevant, and found that its source code is obfuscated.
As such it is automatically classified as malware and I strongly advise you do not use it.

**Comment by mcmonkey4eva** [comment]
I think based on the above (unable to replicate) this issue constitutes user error.

If you can replicate this on a clean server with current builds and no unrelated plugins, please post the full list of everything you've done on such a clean server to replicate it (config settings changed, commands used to replicate it.
Feel free to use an additional bare-minimum-touches-nothing plugin to output event damage values.
Can also use Denizen if you like. I'm using it to test event damages, via this script:
```yml
test:
    type: world
    events:
        on entity damaged:
        - announce "<context.entity.name> hurt for <context.damage> -- <context.final_damage> -- cancelled is <context.cancelled>"
```

Other than that, I would suggest trying to figure out what plugins or world settings (like gamerules) you have that might mess with combat functionality. Also consider trying the steps at https://wiki.citizensnpcs.co/Plugin_Conflict

**Comment by Ferocimo** [comment]
@mcmonkey4eva Okay so, we updated stuff, changed some configs, etc. Results : the issue I was reporting before isn't there anymore. But we have conflict problems with OldCombatMechanics, another plugin we use.

This plugin simulates pre-1.9 PVP. It works fine with PvP, and previously it worked fine with Sentinels. However, now it seems like it's not working properly anymore : 
Enforce damage: true ; workaround damage: true -> NPCs deal a lot of damage, but relatively normally.
Enforce damage: false ; workaround damage: true -> Same thing, no difference at all.
Enforce damage: false ; workaround damage: false -> NPCs deal insane amount of damage, far above what they should do.
Enforce damage true ; workaround damage: false -> NPCs deal almost no damage (about 0.1), whatever we set as damage value. Even if the Sentinel has a damage value of 1000, it will deal 0.1 damage.

Now, we could use workaround damage to make it work, but we would have to reconfigure hundreds of NPCs, and since it could be fixed at any time it would maybe be for nothing...
